optaplanner-core/src/test/java/org/optaplanner/core/api/solver/SolverManagerTest.java

```diff
@@ -240,4 +245,87 @@ public void skipAhead() throws ExecutionException, InterruptedException {
+        solverManager.solve(secondProblemId, PlannerTestUtils.generateTestdataSolution("s2"));
+        assertSame(SolverStatus.SOLVING_SCHEDULED, solverManager.getSolverStatus(secondProblemId));
+
+        // Terminate second job which is in scheduled status. The status changes from SOLVING_SCHEDULED to NOT_SOLVING
+        solverManager.terminateEarly(secondProblemId);
+        waitForJobStatusChangeOrTimeout(solverManager, secondProblemId, SolverStatus.NOT_SOLVING);
+
+        // Reschedule second job. It reintroduces it to the solverManager by changing status from NOT_SOLVING to
+        // SOLVING_SCHEDULED in which it stays.
+        solverManager.solve(secondProblemId, PlannerTestUtils.generateTestdataSolution("s2"));
+        waitForJobStatusChangeOrTimeout(solverManager, secondProblemId, SolverStatus.SOLVING_SCHEDULED);
+
+        // Terminate the first job which triggers it to change it's status from SOLVING_ACTIVE to NOT_SOLVING.
+        // Automatically starts working on the second job. Its status changes from STATUS_SCHEDULED to STATUS_ACTIVE.
+        solverManager.terminateEarly(firstProblemId);
+        waitForJobStatusChangeOrTimeout(solverManager, firstProblemId, SolverStatus.NOT_SOLVING);
+        waitForJobStatusChangeOrTimeout(solverManager, secondProblemId, SolverStatus.SOLVING_ACTIVE);
+
+        // Terminate the second job, its status changes from SOLVING_ACTIVE to NOT_SOLVING. Both jobs stopped solving.
+        solverManager.terminateEarly(secondProblemId);
+        waitForJobStatusChangeOrTimeout(solverManager, secondProblemId, SolverStatus.NOT_SOLVING);
+        assertSame(SolverStatus.NOT_SOLVING, solverManager.getSolverStatus(firstProblemId));
+
+        // Reintroduce the problems to solverManager and attempt to close them all at once using solverManager.close().
+        // The first job's status changes from SOLVING_ACTIVE to NOT_SOLVING.
+        // The second job's status changes from SOLVING_SCHEDULED to NOT_SOLVING.
+        solverManager.solve(firstProblemId, PlannerTestUtils.generateTestdataSolution("s1"));
+        waitForJobStatusChangeOrTimeout(solverManager, firstProblemId, SolverStatus.SOLVING_ACTIVE);
+
+        solverManager.solve(secondProblemId, PlannerTestUtils.generateTestdataSolution("s2"));
+        assertSame(SolverStatus.SOLVING_SCHEDULED, solverManager.getSolverStatus(secondProblemId));
+
+        solverManager.close();
+        waitForJobStatusChangeOrTimeout(solverManager, firstProblemId, SolverStatus.NOT_SOLVING);
+        waitForJobStatusChangeOrTimeout(solverManager, secondProblemId, SolverStatus.NOT_SOLVING);
+    }
+
+    private void waitForJobStatusChangeOrTimeout(SolverManager<TestdataSolution, Long> solverManager, Long problemId, SolverStatus expectedStatusChange) throws InterruptedException {
+        long t = System.currentTimeMillis();
+        long end = t + 5000;
```

Comment:
5000 is risky on our slow jenkins machines (or a community user compiling while something else is running on his/her machine).

I recommend just adding a `@Timeout(600_000)` on the test itself and do

```
while (solverManager.getSolverStatus(problemId) == expectedStatusChange) {
   Thread.sleep(100L);
}
```

Calling currentTimeMillis() every 5 milliseconds might not be as cheap as one would expect. And questionable on Windows that has a 17 millisecond clock granularity :)

Reply:
@rsynek @triceo  That 100 milliseconds, I pulled out of thin air. But I presume michal did the same with the 5 milliseconds. We should figure out a good standard for that number for for such "quick polling" patterns used in tests.

 I wouldn't make it less than 20 milliseconds due to clock granularity on common machines. But 100ms is probably too long, now that I think about it. Having 200 tests that do quick polling with 100ms, means they run at least for 20 seconds.

How about we use 20ms as the quick polling standard? Wdyt?

Reply:
How long does calling a REST method take? That polling should not be able to DDOS a slow rest service.

Reply:
@rsynek These comments didn't make it in (nor have a comment to reject them), I 'll provide a Pr for it together with the issue itself.